app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,471 @@
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_STOP_SERVER;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_TURN_OFF_AFTER_O;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_TURN_ON_AFTER_O;
+
+public class ZimHostActivity extends BaseActivity implements
+    ZimHostCallbacks, ZimHostContract.View, LocationCallbacks {
+
+  @BindView(R.id.startServerButton)
+  Button startServerButton;
+  @BindView(R.id.server_textView)
+  TextView serverTextView;
+  @BindView(R.id.recycler_view_zim_host)
+  RecyclerView recyclerViewZimHost;
+
+  @Inject
+  ZimHostContract.Presenter presenter;
+
+  @Inject
+  AlertDialogShower alertDialogShower;
+
+  private static final String TAG = "ZimHostActivity";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  private static final int LOCATION_SETTINGS_PERMISSION_RESULT = 101;
+  public static final String SELECTED_ZIM_PATHS_KEY = "selected_zim_paths";
+  private static final String IP_STATE_KEY = "ip_state_key";
+  private Task<LocationSettingsResponse> task;
+  ProgressDialog progressDialog;
+
+  private BooksOnDiskAdapter booksAdapter;
+  BookOnDiskDelegate.BookDelegate bookDelegate;
+  HotspotService hotspotService;
+  private String ip;
+  private ServiceConnection serviceConnection;
+  LocationServicesHelper locationServicesHelper;
+  ArrayList<String> selectedBooksPath = new ArrayList<>();
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    setUpToolbar();
+
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(IP_STATE_KEY);
+      layoutServerStarted();
+    }
+    bookDelegate =
+        new BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil,
+            null,
+            null,
+            bookOnDiskItem -> {
+              select(bookOnDiskItem);
+              return Unit.INSTANCE;
+            });
+    bookDelegate.setSelectionMode(SelectionMode.MULTI);
+    booksAdapter = new BooksOnDiskAdapter(bookDelegate,
+        BookOnDiskDelegate.LanguageDelegate.INSTANCE
+    );
+
+    presenter.attachView(this);
+
+    presenter.loadBooks();
+    recyclerViewZimHost.setAdapter(booksAdapter);
+
+    serviceConnection = new ServiceConnection() {
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        HotspotService.HotspotBinder binder = (HotspotService.HotspotBinder) service;
+        hotspotService = binder.getService();
+        hotspotService.registerCallBack(ZimHostActivity.this);
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName arg0) {
+      }
+    };
+
+    startServerButton.setOnClickListener(v -> {
+      //Get the path of ZIMs user has selected
+      if (!ServerUtils.isServerStarted) {
+        getSelectedBooksPath();
```

Comment:
make `getSelectedBookPaths` actually return the selected books and then also probably delete the field `selectedBookPaths`. You won't have to clear the list this way

Reply:
Fixed. 992ce4ad03fb26b917356188349bbe82b75aac19